### PR TITLE
Adding GcStatusMonitor to log Major GCs

### PR DESCRIPTION
**Pull request by aweisberg**
Replaces #13495 which I can't seem to push to anymore.

Please make sure your submission complies with our [Development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [Formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), and [Commit Message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests) guidelines.

Fill in the release notes towards the bottom of the PR description.
See [Release Notes Guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) for details.

```
== RELEASE NOTES ==

General Changes
* ...
* ...

Hive Changes
* ...
* ...
```

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by aweisberg** [comment]
Tests are broken because registering a listener is broken in some tests.

File: presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java [hunk 1]
```diff
@@ -44,7 +44,7 @@
 import static java.util.Collections.nCopies;
 import static java.util.Objects.requireNonNull;
 
-public class WindowOperator
+public class  WindowOperator
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java]: nit: unnecessary change

File: presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+public class GcStatusMonitor
+{
+    private static final Logger log = Logger.get(GcStatusMonitor.class);
+    private static final String GC_NOTIFICATION_TYPE = "com.sun.management.gc.notification";
+    private final NotificationListener notificationListener = (notification, ignored) -> onNotification(notification);
+    private final SqlTaskManager sqlTaskManager;
+
+    @Inject
+    public GcStatusMonitor(
+            SqlTaskManager sqlTaskManager)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java]: nit: unnecessary wrap

File: presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+        long userMemoryReservation = tasks.stream()
+                .map(SqlTask::getTaskInfo)
+                .map(TaskInfo::getStats)
+                .map(TaskStats::getUserMemoryReservation)
+                .mapToLong(DataSize::toBytes)
+                .sum();
+        long systemMemoryReservation = tasks.stream()
+                .map(SqlTask::getTaskInfo)
+                .map(TaskInfo::getStats)
+                .map(TaskStats::getUserMemoryReservation)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java]: `getSystemMemoryReservation`

File: presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+    private void logActiveTasks()
+    {
+        // We only care about active tasks
+        List<SqlTask> activeSqlTasks = getActiveSqlTasks();
+        ListMultimap<QueryId, SqlTask> activeQueriesToTasksMap = getQueriesToTaskMap(activeSqlTasks);
+        logFormatted(activeQueriesToTasksMap);
+    }
+
+    private static ListMultimap<QueryId, SqlTask> getQueriesToTaskMap(List<SqlTask> sqlTasks)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java]: nit: inline?

File: presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+    }
+
+    private List<SqlTask> getActiveSqlTasks()
+    {
+        return sqlTaskManager.getAllTasks().stream()
+                .filter(task -> !task.getTaskInfo().getTaskStatus().getState().isDone())
+                .collect(toImmutableList());
+    }
+
+    private static void logInfoTable(List<List<String>> table)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java]: nit: could you please reorder the methods as they are used?

File: presto-main/src/test/java/com/facebook/presto/server/TestServer.java [hunk 6]
```diff
@@ -85,6 +89,38 @@ public void teardown()
...
+    @Test(timeOut = 10000)
+    public void testServerStartsWithGC()
+            throws Exception
+    {
+        // force a GC
+        long gcCount = getNumGcs();
+        System.gc();
+        while (gcCount >= getNumGcs()) {
+            Thread.sleep(1000);
+            System.gc();
```
Comment (arhimondr) [inline presto-main/src/test/java/com/facebook/presto/server/TestServer.java]: This may or may not cause the full gc. There are JVM flags that can turn it to only trigger a concurrent collection, or even completely ignored. I would simply remove this tests, as effectively it just verifies the MX bean correctness.

File: presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+                    new SimpleEntry<>(entry.getKey(), entry.getValue().stream()
+                            .map(SqlTask::getTaskInfo)
+                            .map(TaskInfo::getStats)
+                            .mapToLong(stats -> stats.getUserMemoryReservation().toBytes() + stats.getSystemMemoryReservation().toBytes())
+                            .sum())
+                ).sorted(comparator.reversed())
+                .map(Entry::getKey)
+                .collect(toImmutableList());
+        queriesByMemoryUsage.stream()
+                .forEach(queryId -> logFormatted(queryId, queryIDToSqlTaskMap.get(queryId)));
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java]: do we want to have one table per query? Or should it be a single table that contains multiple queries?

File: presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+        queriesByMemoryUsage.stream()
+                .map(queryIDToSqlTaskMap::get)
+                .forEach(GcStatusMonitor::logTaskInfos);
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java]: Same question here?

File: presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+                "Input Row Count",
+                "Output Row Count"));
+
+        Comparator<SqlTask> comparator = (a, b) -> {
+            TaskStats aTaskStats = a.getTaskInfo().getStats();
+            TaskStats bTaskStats = b.getTaskInfo().getStats();
+            return Long.compare(aTaskStats.getUserMemoryReservation().toBytes() + aTaskStats.getSystemMemoryReservation().toBytes(),
+                    bTaskStats.getUserMemoryReservation().toBytes() + bTaskStats.getSystemMemoryReservation().toBytes());
+        };
+        sqlTasks.stream()
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java]: It is not added to the `taskLogInfoList`

File: presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+                ImmutableList.of(
+                        queryId.toString(),
+                        Long.toString(userMemoryReservation + systemMemoryReservation),
+                        Long.toString(userMemoryReservation),
+                        Long.toString(systemMemoryReservation))));
+    }
+
+    private static void logTaskInfos(Collection<SqlTask> sqlTasks)
+    {
+        ArrayList<List<String>> taskLogInfoList = new ArrayList<>();
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java]: nit: you can use the ImmutableList.builder() instead

File: presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+                "State",
+                "Created Ts",
+                "User Memory",
+                "System Memory",
+                "Input Bytes",
+                "Output Bytes",
+                "Input Row Count",
+                "Output Row Count"));
+
+        Comparator<SqlTask> comparator = (a, b) -> {
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java]: nit: `first`, `second`?

File: presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+            TaskStats aTaskStats = a.getTaskInfo().getStats();
+            TaskStats bTaskStats = b.getTaskInfo().getStats();
+            return Long.compare(aTaskStats.getUserMemoryReservation().toBytes() + aTaskStats.getSystemMemoryReservation().toBytes(),
+                    bTaskStats.getUserMemoryReservation().toBytes() + bTaskStats.getSystemMemoryReservation().toBytes());
+        };
+        sqlTasks.stream()
+                .sorted(comparator.reversed())
+                .map(task -> {
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java]: The formatting seems to be broken

File: presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+        }).collect(toImmutableList());
+
+        logInfoTable(taskLogInfoList);
+    }
+
+    private static void logFormatted(ListMultimap<QueryId, SqlTask> queryIDToSqlTaskMap)
+    {
+        Comparator<Entry<QueryId, Long>> comparator = Comparator.comparingLong(Entry::getValue);
+        //Print summary of queries sorted from most memory hungry to least memory hungry
+        List<QueryId> queriesByMemoryUsage = queryIDToSqlTaskMap.asMap().entrySet().stream()
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java]: queryId

File: presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+            if (info.isMajorGc()) {
+                onMajorGc();
+            }
+        }
+    }
+
+    private void onMajorGc()
+    {
+        try {
+            logActiveTasks();
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java]: Think to think about: do we think it is fine to do the logging in the notification thread?

Comment (aweisberg) [reply presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java]: Java has a service thread that calls this. The existing airlift GC status monitor is also logging from this and it doesn't seem to be causing issues.

File: presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java (hunk 5 above)
Comment (aweisberg) [reply presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java]: Why are you commenting on logInfoTable in particular? This is a method that is defined before it is ever used.

If I take this to it's logical conclusion it means that the methods that are lowest level and least interesting go first at the top of the file? I can reorder the file this way and push the highest level methods to the bottom and the lowest level methods to the top, but this isn't what I see when I look at the code base. Quite the opposite really. 

I don't care at all. I just want clarity so I can get this done.

File: presto-main/src/test/java/com/facebook/presto/server/TestServer.java (hunk 6 above)
Comment (aweisberg) [reply presto-main/src/test/java/com/facebook/presto/server/TestServer.java]: I don't know how this change is here. It was removed from the original PR.

File: presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java (hunk 13 above)
Comment (aweisberg) [reply presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java]: Renamed to queryIdsSortedByMemoryUsage

File: presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java (hunk 5 above)
Comment (aweisberg) [reply presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java]: Did the reordering.

File: presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java (hunk 4 above)
Comment (aweisberg) [reply presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java]: Inlined.

**Comment by aweisberg** [comment]
Looks like it is working now. Sorry I added you as a reviewer when I created the new PR when it wasn't really done.

I only want to create the header once or each type (query, task). 

I think it's probably fine to log from the GC thread. It's what the existing airlift GC monitor is doing. Adding a thread to process the notifications is also easy enough.

```
2019-12-08T14:49:22.536-0600	INFO	Service Thread	com.facebook.airlift.stats.JmxGcMonitor	Major GC: application 109ms, stopped 202ms: 490.21MB -> 490.03MB
2019-12-08T14:49:22.538-0600	INFO	Service Thread	com.facebook.presto.util.GcStatusMonitor	| Query ID                    | Total Memory Reservation | User Memory Reservation | System Memory Reservation |
2019-12-08T14:49:22.538-0600	INFO	Service Thread	com.facebook.presto.util.GcStatusMonitor	| 20191208_204915_00000_adffw | 0                        | 0                       | 0                         |
2019-12-08T14:49:22.541-0600	INFO	Service Thread	com.facebook.presto.util.GcStatusMonitor	| Query ID                    | Task ID                           | State   | Created Ts                    | User Memory | System Memory | Input Bytes | Output Bytes | Input Row Count | Output Row Count |
2019-12-08T14:49:22.541-0600	INFO	Service Thread	com.facebook.presto.util.GcStatusMonitor	| 20191208_204915_00000_adffw | 20191208_204915_00000_adffw.2.0.0 | RUNNING | 2019-12-08T14:49:19.317-06:00 | 0B          | 0B            | 0           | 0            | 0               | 0                |
2019-12-08T14:49:22.541-0600	INFO	Service Thread	com.facebook.presto.util.GcStatusMonitor	| 20191208_204915_00000_adffw | 20191208_204915_00000_adffw.1.0.0 | RUNNING | 2019-12-08T14:49:19.757-06:00 | 0B          | 0B            | 0           | 0            | 0               | 0                |
2019-12-08T14:49:22.541-0600	INFO	Service Thread	com.facebook.presto.util.GcStatusMonitor	| 20191208_204915_00000_adffw | 20191208_204915_00000_adffw.0.0.0 | RUNNING | 2019-12-08T14:49:19.773-06:00 | 0B          | 0B            | 0           | 0            | 0               | 0                |
```

File: presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java (hunk 5 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java]: > Why are you commenting on logInfoTable in particular? This is a method that is defined before it is ever used.

Sorry, it was for no reason. Just a random line in the file.

>  I can reorder the file this way and push the highest level methods to the bottom and the lowest level methods to the top, but this isn't what I see when I look at the code base.

All across the codebase (generally) we are trying to maintain the highest level methods -> lowest level methods.

e.g.: When A calls B and B calls C the methods should be ordered as 

```
A {
B()
}

B {
C()
}

C {
...
}

```

So when you are reading a class upside down - you start with A. If A is not enough, and more details are needed - you will probably be looking for B, and the most intuitive place to find it  - is right below A. And so on ... 


I'm not sure if there's a strong guideline on this. The Google java guideline basically says

> What is important is that each class uses some logical order, which  ...

https://google.github.io/styleguide/javaguide.html#s3.4.2-class-member-ordering

However all across the code generally we were trying to maintain the `highest level methods -> lowest level methods`. 

Perhaps we should update our coding guidelines to reflect this implicit convention

**Comment by arhimondr** [comment]
Please squash the commits before merging

**Comment by aweisberg** [comment]
Squashed

File: presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java (hunk 5 above)
Comment (aweisberg) [reply presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java]: Heh, OK that is the ordering I expect. The previous PR had me confused. I'll reorder it.
